presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantCastToVarcharInJoinClause.java

```diff
@@ -0,0 +1,179 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.common.type.VarcharType;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.metadata.CastType;
+import com.facebook.presto.metadata.FunctionAndTypeManager;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.ProjectNode;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.relational.FunctionResolution;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.SystemSessionProperties.isRemoveRedundantCastToVarcharInJoinEnabled;
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.IntegerType.INTEGER;
+import static com.facebook.presto.sql.planner.PlannerUtils.addProjections;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.relational.Expressions.call;
+
+/**
+ * Remove redundant cast to varchar in join condition for queries like `select select * from orders o join customer c on cast(o.custkey as varchar) = cast(c.custkey as varchar)`
+ * Transform from
+ * <pre>
+ *     - Join
+ *          left_cast = right_cast
+ *          - Project
+ *              left_cast := cast(lkey as varchar)
+ *              - TableScan
+ *                  lkey BIGINT
+ *          - Project
+ *              right_cast := cast(rkey as varchar)
+ *              - TableScan
+ *                  rkey BIGINT
+ *
+ * </pre>
+ * into
+ * <pre>
+ *     - Join
+ *          new_lkey = new_rkey
+ *          - Project
+ *              left_cast := cast(lkey as varchar)
+ *              new_lkey := lkey
+ *              - TableScan
+ *                  lkey BIGINT
+ *          - Project
+ *              right_cast := cast(rkey as varchar)
+ *              new_rkey := rkey
+ *              - TableScan
+ *                  rkey BIGINT
+ * </pre>
+ * We will rely on optimizations later to remove unnecessary cast (if not used) and identity projection here.
+ *
+ * Notice that we do not apply similar optimizations to queries with similar join condition like `cast(bigint as varchar) = varchar`. In general it can be converted to
```

Comment:
should we allow this transformation for all non-lossy/deterministic casts?

Reply:
I applied for cast(bigint/integer as varchar) as varchar has high cost for hash operation. Other cases may not show as big improvement here.